Remove annotation support

Spea opened the pull request:
Drops support for serializer annotations. Type annotations still have to be parsed in order to properly set the types in the metadata.

dbu commented on src/Annotation/Preferred.php:
we should have made extending classes in the Attribute namespace rather than just make the annotations attributes too, so that people switching to attributes would have used the right namespace from the beginning, but aparently we did not...

imho now is the last chance to somewhat cleanly fix the namespace. people might still be on attributes and then they have to change anyways when upgrading. and people already on attributes will need to adjust their `use` statements.

Spea replied:
Yeah makes sense to avoid confusion here. I moved the class to the `Attribute` namespace now.

dbu commented on src/Exception/ParseException.php:
```diff
@@ -10,10 +10,10 @@ final class ParseException extends SchemaException
     private const PROPERTY_ERROR = 'Property "%s::%s" couldn\'t be parsed because of an error: %s';
     private const PROPERTY_TYPE_ERROR = 'Property "%s::%s" has an invalid type which results in: %s';
-    private const PROPERTY_TYPE_NAME_NULL = 'Property "%s::%s" has an invalid type annotation. [Type Error] Attribute "name" of @JMS\Type may not be null.';
+    private const PROPERTY_TYPE_NAME_NULL = 'Property "%s::%s" has an invalid type attribute. [Type Error] Attribute "name" of @JMS\Type may not be null.';
```
```suggestion
    private const PROPERTY_TYPE_NAME_NULL = 'Property "%s::%s" has an invalid type attribute. [Type Error] Attribute "name" of #JMS\Type may not be null.';
```
should we change the @ to # to match that we talk about attribute and not annotation?

Spea replied:
I agree and I adjusted it accordingly.

dbu commented on src/ModelParser/LiipMetadataAnnotationParser.php:
```diff
@@ -4,30 +4,21 @@
  * this parser runs. No error is raised if a field is marked as preferred but
  * not included in the meta data.
  */
 final class LiipMetadataAnnotationParser implements ModelParserInterface
```
do we still respect any annotations as well? if not, we shoudl rename this class too

Spea replied:
We do not, I changed the name to `LiipMetadataAttributeParser`

dbu reviewed (commented):
thanks for this cleanup!

i see one problem with this: jms serializer still supports annotations. it might confuse people massively when suddenly their annotations are ignored. i wonder if we need to keep reading the jms annotations if present, and only drop our own annotations support  (which should error with properly configured php, when trying to use a no longer existing annotation)

Spea committed:
Change wording to `attribute(s)`

Support for JMS annotations will be kept for now (until the library itself drops the support), but since attributes is a known wording in the PHP world now, we change `annotation(s)` to `attribute(s)`

Spea committed:
Remove annotation support for `Preferred` attribute

We no longer want to support annotations for the `Preferred` class, so we adjust the namespace and naming of the parser itself.

Spea commented:
> thanks for this cleanup!
> 
> i see one problem with this: jms serializer still supports annotations. it might confuse people massively when suddenly their annotations are ignored. i wonder if we need to keep reading the jms annotations if present, and only drop our own annotations support (which should error with properly configured php, when trying to use a no longer existing annotation)

I was already hesitant in removing annotation support for JMS anyway, so I agree that we should keep it (for now) and only remove it if JMS itself decides to remove it as well. I reverted the changes that removed the annotation support, but I still went with renaming `annotation(s)` to `attribute(s)`.

dbu commented on CHANGELOG.md:
```diff
@@ -12,6 +12,7 @@
 * Adjusted code to not trigger warnings with PHP 8.4
 * Removed deprecated `getDeserializeFormat` method from date time
 * Add attribute support for the `@Preferred` annotation 
+* Remove annotation support for the `Preferred` class and only allow using it as an attribute
```
we should merge this with the above line: 
```
* Replaced `@Preferred` annotation with the `#Preferred` attribute
```

Spea replied:
Yeah makes sense to not have two entries for the same change. I fixed this and amended to the original commit.

dbu reviewed (commented):
thanks a lot. yes i agree with the renaming, that is a potential BC break so we need to do this now. 
when jms ditches annotations, we can do another major version.

in #48 , we had the idea to make attributes optional and not require the library anymore. then we would have to change the attribute parser to not attempt to read annotations if the library is not available. if you are up for that, would be cool to change it. but i am fine keeping it if you don't have the time to invest into it.

Spea commented:
> in #48 , we had the idea to make attributes optional and not require the library anymore. then we would have to change the attribute parser to not attempt to read annotations if the library is not available. if you are up for that, would be cool to change it. but i am fine keeping it if you don't have the time to invest into it.

I'll gladly do this, but if its fine for you I'd rather do that in a separate PR.

dbu reviewed (approved):
yeah agree with separate merge requests.

thanks!

Spea commented:
> > in #48 , we had the idea to make attributes optional and not require the library anymore. then we would have to change the attribute parser to not attempt to read annotations if the library is not available. if you are up for that, would be cool to change it. but i am fine keeping it if you don't have the time to invest into it.
> 
> I'll gladly do this, but if its fine for you I'd rather do that in a separate PR.

@dbu I was thinking about this a bit more, and I wonder if we can actually do this, since there is still the [PhpTypeParser](https://github.com/liip/metadata-parser/blob/b2897803104cc399b317c0a685d586d14e10bd78/src/TypeParser/PhpTypeParser.php#L16) that takes care of checking the types.

dbu commented:
oh, you are very right. and i  don't see us dropping the php type annotations anytime soon. afaik there is still no other way to e.g. do typed arrays.

then i think we don't need to do anything more on that for now.

from my point of view, this would be ready to tag a release. should i go for it or did you want to do something else before i tag?

Spea commented:
> from my point of view, this would be ready to tag a release. should i go for it or did you want to do something else before i tag?

Fine for me. I'm still working on the discriminator support, but that might take a bit longer. Though nothing is pressuring us since the serializer also has to go to a new major version when using the metadataparser `2.x`, correct?

Anyway, I think its fine to do a release and we add discriminator support in `2.1` then.
